causallearn/utils/cit.py

```diff
@@ -15,78 +15,142 @@
+        Xs: List<int>, sorted. may swapped with Ys for cache key uniqueness.
+        Ys: List<int>, sorted.
+        condition_set: List<int>
+        cache_key: string. Unique for <X,Y|S> in any input type or order.
+        '''
+        def _stringize(ulist1, ulist2, clist):
+            # ulist1, ulist2, clist: list of ints, sorted.
+            _strlst  = lambda lst: '.'.join(map(str, lst))
+            return f'{_strlst(ulist1)};{_strlst(ulist2)}|{_strlst(clist)}' if len(clist) > 0 else \
+                   f'{_strlst(ulist1)};{_strlst(ulist2)}'
+
+        # every time when cit is called, auto save to local cache.
+        self.save_to_local_cache()
+
+        METHODS_SUPPORTING_MULTIDIM_DATA = ["kci"]
+        if condition_set is None: condition_set = []
+        # 'int' to convert np.*int* to built-in int; 'set' to remove duplicates; sorted for hashing
+        condition_set = sorted(set(map(int, condition_set)))
+
+        # usually, X and Y are 1-dimensional index (in constraint-based methods)
+        if self.method not in METHODS_SUPPORTING_MULTIDIM_DATA:
+            X, Y = (int(X), int(Y)) if (X < Y) else (int(Y), int(X))
+            assert X not in condition_set and Y not in condition_set, "X, Y cannot be in condition_set."
+            return [X], [Y], condition_set, _stringize([X], [Y], condition_set)
+
+        # also to support multi-dimensional unconditional X, Y (usually in kernel-based tests)
+        Xs = sorted(set(map(int, X))) if isinstance(X, Iterable) else [int(X)]  # sorted for comparison
+        Ys = sorted(set(map(int, Y))) if isinstance(Y, Iterable) else [int(Y)]
+        Xs, Ys = (Xs, Ys) if (Xs < Ys) else (Ys, Xs)
+        assert len(set(Xs).intersection(condition_set)) == 0 and \
+               len(set(Ys).intersection(condition_set)) == 0, "X, Y cannot be in condition_set."
+        return Xs, Ys, condition_set, _stringize(Xs, Ys, condition_set)
+
+class FisherZ(CIT_Base):
+    def __init__(self, data, **kwargs):
+        super().__init__(data, **kwargs)
+        self.check_cache_method_consistent('fisherz', -1)   # -1: no parameters can be specified for fisherz
+        self.assert_input_data_is_valid()
+        self.correlation_matrix = np.corrcoef(data.T)
+
+    def __call__(self, X, Y, condition_set=None):
+        '''
+        Perform an independence test using Fisher-Z's test.
 
         Parameters
         ----------
-        data : data matrices
         X, Y and condition_set : column indices of data
 
         Returns
         -------
         p : the p-value of the test
-        """
-        var = list((X, Y) + condition_set)
+        '''
+        Xs, Ys, condition_set, cache_key = self.get_formatted_XYZ_and_cachekey(X, Y, condition_set)
+        if cache_key in self.pvalue_cache: return self.pvalue_cache[cache_key]
+        var = Xs + Ys + condition_set
         sub_corr_matrix = self.correlation_matrix[np.ix_(var, var)]
         try:
             inv = np.linalg.inv(sub_corr_matrix)
```

Comment:
hmmm, curious, what if exception is thrown here? You didn't catch?

Reply:
Oh yes I didn't catch exceptions. But what would be the possible exceptions here? If it's about the type of X, Y, condition_set, a built-in error message seems to be informative enough.

Reply:
why do you "try" then?

Reply:
Oh you mean which lines? L155?

This is from the original code at [Fixed fisherz test](https://github.com/cmu-phil/causal-learn/commit/dd1a13edcadc3b50baa51053c6e91f702bbf06e3) (#58).

Reply:
ha, ic. I missed the later "except".... nvm